File: packages/runtime/src/direct-transport.ts
// In-process transport — hands a `CallRequest` straight to the host with no encode/decode hop
// (the generated client already encoded the args; the host's dispatch decodes them). Useful for
// tests and for embedding the host in the same runtime as the client (SSR). Network transports
// are user-supplied.
//
// The forwarded context is caller-supplied and DEFAULT NON-SYSTEM, so visibility gating is
// exercised; `isSystem: true` opts into the in-process system identity (bypasses gating) for SSR.

import type {
    CallRequest, CallResult, RequestContext, Transport, TransportCapabilities, WireEncoding,
} from "./types.js";
import type { ServiceHost } from "./service-host.js";

const NO_STREAMING: TransportCapabilities = { serverStream: false, clientStream: false, bidi: false };

export type DirectTransportOptions = {
    /** Wire encoding a bound client marshals with. Default `"json"`. */
    encoding?: WireEncoding;
    /** Forward the context as the in-process system identity (bypasses visibility gating).
     *  Default false, so gating is exercised; opt in for trusted SSR. */
    isSystem?: boolean;
    /** Per-call context factory (e.g. request-scoped auth). */
    context?: () => RequestContext | Promise<RequestContext>;
};

export function createDirectTransport(host: ServiceHost, options: DirectTransportOptions = {}): Transport {
    const encoding: WireEncoding = options.encoding ?? "json";
    return {
        encoding,
        capabilities: NO_STREAMING,
        async invoke(request: CallRequest): Promise<CallResult> {
            const base: RequestContext = options.context ? await options.context() : {};
            const ctx: RequestContext = options.isSystem
                ? { ...base, identity: { ...(base.identity ?? {}), isSystem: true } }
                : base;
            return host.invoke(request, ctx, encoding);
        },
    };
}

File: packages/runtime-python/src/keyma/runtime/direct_transport.py
"""The in-process direct transport.

``create_direct_transport`` hands a :class:`CallRequest` straight to a :class:`ServiceHost` with no
encode/decode hop — the cheap synchronous path (it ``await``s the host inline). It forwards a
caller-supplied ``ctx``, **default non-system** so the host's visibility gate is genuinely
exercised; pass ``is_system=True`` to opt into system identity (e.g. SSR). The configured
``encoding`` is the one both ends agree on statically."""

from __future__ import annotations

from typing import Optional

from .service_host import ServiceHost
from .transport import CallRequest, CallResult, TransportCapabilities
from .types import Encoding, RequestContext


class DirectTransport:
    """A :class:`Transport` bound to an in-process :class:`ServiceHost`."""

    def __init__(self, host: ServiceHost, ctx: RequestContext, encoding: Encoding) -> None:
        self._host = host
        self._ctx = ctx
        self.encoding: Encoding = encoding
        self.capabilities = TransportCapabilities(unary=True)

    async def invoke(self, request: CallRequest) -> CallResult:
        return await self._host.handle(request, self._ctx, self.encoding)


def create_direct_transport(
    host: ServiceHost,
    *,
    ctx: Optional[RequestContext] = None,
    is_system: bool = False,
    encoding: Encoding = "json",
) -> DirectTransport:
    base_ctx: RequestContext = dict(ctx or {})
    if is_system:
        identity = dict(base_ctx.get("identity") or {})
        identity["isSystem"] = True
        base_ctx["identity"] = identity
    return DirectTransport(host, base_ctx, encoding)
